File: third-party/desktop_multi_window/macos/Classes/MultiWindowManager.swift
//
//  MultiWindowManager.swift
//  flutter_multi_window
//
//  Created by Bin Yang on 2022/1/10.
//

import FlutterMacOS
import Foundation

class MultiWindowManager {
  static let shared = MultiWindowManager()

  private var id: Int64 = 0

  private var windows: [Int64: BaseFlutterWindow] = [:]

  func create(arguments: String) -> Int64 {
    id += 1
    let windowId = id

    let window = FlutterWindow(id: windowId, arguments: arguments)
    window.delegate = self
    window.windowChannel.methodHandler = self.handleMethodCall
    windows[windowId] = window
    return windowId
  }

  func attachMainWindow(window: NSWindow, _ channel: WindowChannel) {
    let mainWindow = BaseFlutterWindow(window: window, channel: channel)
    mainWindow.windowChannel.methodHandler = self.handleMethodCall
    windows[0] = mainWindow
  }

  private func handleMethodCall(fromWindowId: Int64, targetWindowId: Int64, method: String, arguments: Any?, result: @escaping FlutterResult) {
    guard let window = self.windows[targetWindowId] else {
      result(FlutterError(code: "-1", message: "failed to find target window. \(targetWindowId)", details: nil))
      return
    }
    window.windowChannel.invokeMethod(fromWindowId: fromWindowId, method: method, arguments: arguments, result: result)
  }

  func show(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.show()
  }

  func hide(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.hide()
  }

  func isHidden(windowId: Int64) -> Bool {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return true
    }
    return window.isHidden()
  }

  func close(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.close()
  }

  func closeAll() {
    windows.forEach { _, value in
      value.close()
    }
  }

  func center(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.center()
  }

  func focus(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.focus()
  }

  func minimize(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.minimize()
  }

  func showTitleBar(windowId: Int64, show: Bool) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.showTitleBar(show: show)
  }

  func maximize(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.maximize()
  }

  func isMaximized(windowId: Int64) -> Bool {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return false
    }
    return window.isMaximized()
  }

  func isMinimized(windowId: Int64) -> Bool {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return false
    }
    return window.isMinimized()
  }

  func startDragging(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.startDragging()
  }

  func startResizing(windowId: Int64, arguments: [String: Any?]) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.startResizing(arguments: arguments)
  }

  func unmaximize(windowId: Int64) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.unmaximize()
  }
    
    func isFullScreen(windowId: Int64) -> Bool {
        guard let window = windows[windowId] else {
          debugPrint("window \(windowId) not exists.")
          return false
        }
        return window.isFullScreen()
    }

  func setFullscreen(windowId: Int64, fullscreen: Bool) {
      guard let window = windows[windowId] else {
        debugPrint("window \(windowId) not exists.")
        return
      }
      window.setFullscreen(fullscreen: fullscreen)
  }

  func setFrame(windowId: Int64, frame: NSRect) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.setFrame(frame: frame)
  }

  func getFrame(windowId: Int64) -> NSDictionary {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
        return [:];
    }
    return window.getFrame()
  }

  func setTitle(windowId: Int64, title: String) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.setTitle(title: title)
  }

  func resizable(windowId: Int64, resizable: Bool) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.resizable(resizable: resizable)
  }

  func setFrameAutosaveName(windowId: Int64, name: String) {
    guard let window = windows[windowId] else {
      debugPrint("window \(windowId) not exists.")
      return
    }
    window.setFrameAutosaveName(name: name)
  }

  func getAllSubWindowIds() -> [Int64] {
    return windows.keys.filter { $0 != 0 }
  }
    
    func isPreventClose(windowId: Int64) -> Bool {
        guard let window = windows[windowId] else {
          debugPrint("window \(windowId) not exists.")
            return false
        }
        return window.isPreventClose()
    }
    
    func setPreventClose(windowId: Int64, setPreventClose: Bool) {
        guard let window = windows[windowId] else {
          debugPrint("window \(windowId) not exists.")
            return
        }
        return window.setPreventClose(setPreventClose: setPreventClose)
    }
}

protocol WindowManagerDelegate: AnyObject {
  func onClose(windowId: Int64)
}

extension MultiWindowManager: WindowManagerDelegate {
  func onClose(windowId: Int64) {
    // notify flutter to close this window gracefully
    guard let window = self.windows[windowId] else {
      return
    }
    let args = NSDictionary()
    debugPrint("invoke onDestroy \(windowId)")
    window.windowChannel.invokeMethod(fromWindowId: 0, method: "onDestroy", arguments: args, result: nil)
    windows.removeValue(forKey: windowId)
  }
}

File: third-party/desktop_multi_window/linux/multi_window_manager.cc
//
// Created by yangbin on 2022/1/11.
//

#include "multi_window_manager.h"

#include <iostream>
#define RLOCK_WINDOW pthread_rwlock_rdlock(this->get_map_lock())
#define WLOCK_WINDOW pthread_rwlock_wrlock(this->get_map_lock())
#define UNLOCK_WINDOW pthread_rwlock_unlock(this->get_map_lock())

namespace {
int64_t g_next_id_ = 0;

class FlutterMainWindow : public BaseFlutterWindow {

 public:

  FlutterMainWindow(GtkWidget *window, std::unique_ptr<WindowChannel> window_channel)
      : window_channel_(std::move(window_channel)), window_(window) {}

  WindowChannel *GetWindowChannel() override {
    return window_channel_.get();
  }
 protected:
  GtkWindow *GetWindow() override {
    return GTK_WINDOW(window_);
  }

 private:
  std::unique_ptr<WindowChannel> window_channel_;
  GtkWidget *window_;

};

}

// static
MultiWindowManager *MultiWindowManager::Instance() {
  static auto manager = std::make_shared<MultiWindowManager>();
  return manager.get();
}

MultiWindowManager::MultiWindowManager() : windows_() {
  this->windows_map_lock_ = PTHREAD_RWLOCK_INITIALIZER;
}

MultiWindowManager::~MultiWindowManager() = default;

int64_t MultiWindowManager::Create(const std::string &args) {
  g_next_id_++;
  int64_t id = g_next_id_;
  auto window = std::make_unique<FlutterWindow>(id, args, shared_from_this());
  window->GetWindowChannel()->SetMethodHandler([this](int64_t from_window_id,
                                                      int64_t target_window_id,
                                                      const gchar *method,
                                                      FlValue *arguments,
                                                      FlMethodCall *method_call) {
    HandleMethodCall(from_window_id, target_window_id, method, arguments, method_call);
  });
  WLOCK_WINDOW;
  windows_[id] = std::move(window);
  UNLOCK_WINDOW;
  return id;
}

void MultiWindowManager::AttachMainWindow(GtkWidget *main_flutter_window,
                                          std::unique_ptr<WindowChannel> window_channel) {
  RLOCK_WINDOW;                                          
  if (windows_.count(0) != 0) {
    g_critical("AttachMainWindow : main window already exists.");
    UNLOCK_WINDOW;
    return;
  }
  UNLOCK_WINDOW;
  window_channel->SetMethodHandler([this](int64_t from_window_id,
                                          int64_t target_window_id,
                                          const gchar *method,
                                          FlValue *arguments,
                                          FlMethodCall *method_call) {
    HandleMethodCall(from_window_id, target_window_id, method, arguments, method_call);
  });
  WLOCK_WINDOW;
  windows_[0] = std::make_unique<FlutterMainWindow>(main_flutter_window, std::move(window_channel));
  UNLOCK_WINDOW;
}

void MultiWindowManager::HandleMethodCall(int64_t from_window_id,
                                          int64_t target_window_id,
                                          const gchar *method,
                                          FlValue *arguments,
                                          FlMethodCall *method_call
) {
  RLOCK_WINDOW;
  if (windows_.count(target_window_id) == 0) {
    fl_method_call_respond_error(method_call, "-1", "target window not found.", nullptr, nullptr);
    UNLOCK_WINDOW;
    return;
  }
  UNLOCK_WINDOW;
  
  RLOCK_WINDOW;
  auto window_channel = windows_[target_window_id]->GetWindowChannel();
  if (!window_channel) {
    fl_method_call_respond_error(method_call, "-1", "target window channel not found.", nullptr, nullptr);
    UNLOCK_WINDOW;
    return;
  }
  window_channel->InvokeMethod(from_window_id, method, arguments, method_call);
  UNLOCK_WINDOW;
}

void MultiWindowManager::Show(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Show();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::Focus(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Focus();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::Hide(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Hide();
  }
  UNLOCK_WINDOW;
}

bool MultiWindowManager::IsHidden(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    bool ret = window->second->IsHidden();
    UNLOCK_WINDOW;
    return ret;
  }
  UNLOCK_WINDOW;
  return false;
}

void MultiWindowManager::Close(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Close();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::SetFullscreen(int64_t id, bool fullscreen) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->SetFullscreen(fullscreen);
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::SetFrame(int64_t id, double x, double y, double width, double height) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->SetBounds(x, y, width, height);
  }
  UNLOCK_WINDOW;
}

FlValue* MultiWindowManager::GetFrame(int64_t id) {
  FlValue* frame = NULL;
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    frame = window->second->GetBounds();
  } else {
    frame = fl_value_new_map();
  }
  UNLOCK_WINDOW;
  return frame;
}

void MultiWindowManager::SetTitle(int64_t id, const std::string &title) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->SetTitle(title);
  }
  UNLOCK_WINDOW;
}

std::vector<int64_t> MultiWindowManager::GetAllSubWindowIds() {
  RLOCK_WINDOW;
  std::vector<int64_t> ids;
  for (auto &window : windows_) {
    if (window.first != 0) {
      ids.push_back(window.first);
    }
  }
  UNLOCK_WINDOW;
  return ids;
}

void MultiWindowManager::Center(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Center();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::StartDragging(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->StartDragging();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::Minimize(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Minimize();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::Maximize(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Maximize();
  }
  UNLOCK_WINDOW;
}

bool MultiWindowManager::IsMaximized(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    UNLOCK_WINDOW;
    return window->second->IsMaximized();
  }
  UNLOCK_WINDOW;
  return false;
}

bool MultiWindowManager::IsMinimized(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    UNLOCK_WINDOW;
    return window->second->IsMinimized();
  }
  UNLOCK_WINDOW;
  return false;
}

void MultiWindowManager::Unmaximize(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->Unmaximize();
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::ShowTitlebar(int64_t id, bool show) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->ShowTitlebar(show);
  }
  UNLOCK_WINDOW;
}

void MultiWindowManager::OnWindowClose(int64_t id) {}

void MultiWindowManager::OnWindowDestroy(int64_t id) {
  std::cout << "destory id " << id << std::endl;
  WLOCK_WINDOW;
  windows_.erase(id);
  UNLOCK_WINDOW;
}

void MultiWindowManager::StartResizing(int64_t id, FlValue *value) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->StartResizing(value);
  }
  UNLOCK_WINDOW;
}

bool MultiWindowManager::IsPreventClose(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    bool ret = window->second->IsPreventClose();
    UNLOCK_WINDOW;
    return ret;
  }
  UNLOCK_WINDOW;
  return false;
}

void MultiWindowManager::SetPreventClose(int64_t id, bool setPreventClose) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    window->second->SetPreventClose(setPreventClose);
  }
  UNLOCK_WINDOW;
}

pthread_rwlock_t* MultiWindowManager::get_map_lock() {
  return &this->windows_map_lock_;
}

int64_t MultiWindowManager::GetXID(int64_t id) {
  RLOCK_WINDOW;
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    auto xid = window->second->GetXID();
    UNLOCK_WINDOW;
    return xid;
  }
  UNLOCK_WINDOW;
  return -1;
}

bool MultiWindowManager::IsFullScreen(int64_t id) {
  auto window = windows_.find(id);
  if (window != windows_.end()) {
    return window->second->IsFullScreen();
  }
  return false;
}